### ISSUE TITLE
MechanicsTensileCT test sensitive to OMP settings

### ISSUE THREAD

**Issue by bartgol** [gahansen/Albany#339]
I was chasing down the issue with a failing test (MechanicsTensileCT), and I finally found the problem: the test passes with OMP_NUM_THREADS=1, but fails otherwise. I am not sure where the problem is, but some of the LCM folks might want to look into this. 

Some extra details. With more than 1 thread, the tests fail due to a failing regression test. Something like

`Response Test 0: 3.066335674548e-05 != 3.068000000000e-05 (rel 1.000000000000e-05 abs 1.000000000000e-08)`

The computed value is deterministic, in the sense that different runs with the same number of threads lead to the same response value. However, different number of threads lead to different response values. This makes me think it is not necessarily a race condition, but I'm not really sure how to explain it otherwise.

Note: this happens on my machine also with master. Compiling with gcc 6.1.0, against Trilinos develop 8e9d4ba . It is not the most recent Trilinos, but I keep 'update trilinos' always as my last resort. Besides, the testsuite is otherwise passing, so I _think_ it is not an issue with Trilinos...

Tagging some LCM people: @ikalash @lxmota

**Comment by bartgol** [comment]
I edited the title of the issue cause I discovered that another failing test was due to the same problem. Namely, CrystalPlasticity_OrientationOnMesh fails on my machine with more than 1 thread, due to an RCP count reaching 0 before the end of the execution.

**Comment by bartgol** [comment]
I believe the problem (at least for CrystalPlasticity_OrientationOnMesh) is that operator() of the Kernel creates an integrator factory which stores an RCP. The creation/destruction of an RCP is _not_ thread safe, so the counter can get really screwed up with more than 1 thread.

@ikalash @lxmota can you confirm this?

By the way, I think the way ParallelConstitutiveModel passes the kernel to the ||for is wrong: i believe this should not work on CUDA. The correct way would be to create a stack copy of the _kernel_ (not its pointer), which will then be copied to the device memory at dispatch time.

**Comment by bartgol** [comment]
Thanks Irina. I am in no rush. Though I will consider my branch mergeable, since I get the same pass/fail as in master, with all tests passing with OMP_NUM_THREADS=1. 

FYI, I haven't tested CUDA, the issue I reported is for OpenMP (though I don't see how that code could work on GPU). Due to the complexity of CUDA builds, I will wait for the nightly to tell me whether that needs to be investigated or not... ;)

**Comment by bartgol** [comment]
It appears the cdash confirms this issue:

[](http://cdash.sandia.gov/CDash-2-3-0/index.php?project=Albany&subproject=IKTAlbanyFunctorOpenMP)

**Comment by lxmota** [comment]
The following test failures may be related to this, as they are all segfaults that have to do with Tpetra and Kokkos:

http://cdash.sandia.gov/CDash-2-3-0/viewTest.php?onlyfailed&buildid=74341
http://cdash.sandia.gov/CDash-2-3-0/viewTest.php?onlyfailed&buildid=74344

**Comment by ikalash** [comment]
@gphlipot : per our discussion this morning, can you please check if the drop test case you are running seg faults in a similar way in the master branch of Albany, and point me to the input files if it does?  I'd like to see if it seg faults on my machine or only on the LCM machines having the very recent gcc version.

**Comment by ikalash** [comment]
I just talked to @lxmota , and he said he will look at this.  Alejandro, here are links to the CDash where you can see the failures:

Blake: https://my.cdash.org/viewTest.php?onlyfailed&buildid=1537370 (ignore the MPS, ATO and Elasticity2DTraction tests).
Camobap: http://cdash.sandia.gov/CDash-2-3-0/viewTest.php?onlyfailed&buildid=76664 (ignore Elasticity2DTracton test).

**Comment by bartgol** [comment]
Yes, IIRC, with OMP_NUM_THREADS=1 the tests would pass. This strongly hints to a race condition somewhere.

**Comment by ikalash** [comment]
Yes I confirmed what you say @bartgol (things are good with OMP_NUM_THREADS=1) when @lxmota and I were looking at it earlier today.

**Comment by ikalash** [comment]
To document a little bit more the behavior: this issue seems to affect the tests

CrystalPlasticity_SchwarzBar_modified_np1
MechanicsTensileCT
CrystalPlasticity_OrientationOnMesh 
CrystalPlasticity_OrientationOnMesh_np4

As Luca mentioned above, they pass with some OMP_NUM_THREADS in an OpenMP build, fail with others.  On my machine camobap, they seg fault, e.g., http://cdash.sandia.gov/CDash-2-3-0/testDetails.php?test=4151443&build=79245.  An interesting thing I noticed is the behavior seems affected also by the --map-by core argument passed to mpirun.  Without this, on my machine the tests run with 2 threads.

@lxmota I know you had the intention to fix this but for the time being I have turned off the tests in the OpenMP builds in an effort to have a clean dashboard.  I can turn them back on once this issue is resolved.  Keeping this issue open for the time being.
